keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
++ Implement a Kubernetes controller architecture with support for pluggable provisioners
++ As MVP, be accessible to the largest groups of consumers by supporting the major object storage protocols (S3, Google Cloud Storage, Azure Blob) while being extensible for future protocol additions.
++ Present similar workflows for both greenfield and brownfield bucket operations.
+
+## Non-Goals
+
++ Define the _data-plane_ object store interface to replace or supplement existing vendor interfaces (i.e. replace GCS, S3, or Azure Blob)
+
+##  Vocabulary
+
++  _Brownfield Bucket_ - a storage instance created out of band.
++ _BucketRequest_ - A user-namespaced custom resource representing a request for a storage instance endpoint.
++  _BucketClass_ - A cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set for creating new buckets
++ _Bucket_ - A cluster-scoped custom resource referenced by a Bucket and containing connection information and metadata for a storage instance.
++ _Greenfield Bucket_ - a bucket created by automation.
++  _Object_ - An atomic, immutable unit of data stored in buckets.
++  *Storage Instance* - Refers to the back object storage endpoint being abstracted by the Bucket API (a.k.a “bucket” or “container”).
++ _Driverless_ - A system where not driver is deployed to automate object store operations.  COSI automation may still be deployed to managed COSI APIs.  See [Architecture](#architecture).
+
+# Proposal
+
+## User Stories
+
+#### Admin
+
+- As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I can see all object storage instances from a single pane, regardless of the backing storage vendor.
+
+#### User
+
+- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
+- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
+- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.
+
+## APIs
+
+> NOTE: For a graphical mapping of how these APIs relate to each other, see [API Relationships](#api-relationships).
+
+
+
+### Storage APIs
+
+#### BucketRequest
+
+A namespaced API representing a workload’s need for a storage instance endpoint. Created in the workload's namespace.  A [BucketAccessRequest](#bucketaccessrequest) is required in order to gain credentialed access to the storage instance.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
```

Comment:
bucketclass is only used for greenfield - optional.

Reply:
@jeffvance This comment is conflicting with the [5] explanation later - which one is more uptodate?

Reply:
We've gone back and forth several times on requiring BC for brownfield. Good catch.